packages/valory/skills/mech_interact_abci/rounds.py

```diff
@@ -0,0 +1,211 @@
+# -*- coding: utf-8 -*-
+# ------------------------------------------------------------------------------
+#
+#   Copyright 2023 Valory AG
+#
+#   Licensed under the Apache License, Version 2.0 (the "License");
+#   you may not use this file except in compliance with the License.
+#   You may obtain a copy of the License at
+#
+#       http://www.apache.org/licenses/LICENSE-2.0
+#
+#   Unless required by applicable law or agreed to in writing, software
+#   distributed under the License is distributed on an "AS IS" BASIS,
+#   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#   See the License for the specific language governing permissions and
+#   limitations under the License.
+#
+# ------------------------------------------------------------------------------
+
+"""This package contains the rounds of MechInteractAbciApp."""
+
+from enum import Enum
+from typing import Dict, List, Optional, Set, Tuple, cast
+
+from packages.valory.skills.abstract_round_abci.base import (
+    AbciApp,
+    AbciAppTransitionFunction,
+    CollectSameUntilThresholdRound,
+    AppState,
+    BaseSynchronizedData,
+    DegenerateRound,
+    EventToTimeout,
+    OnlyKeeperSendsRound,
+    get_name
+)
+
+from packages.valory.skills.mech_interact_abci.payloads import (
+    MechRequestPayload,
+    MechResponsePayload,
+    MechRandomnessPayload,
+    MechSelectKeeperPayload
+)
+
+
+class Event(Enum):
+    """MechInteractAbciApp Events"""
+
+    ROUND_TIMEOUT = "round_timeout"
+    DONE = "done"
+    NO_MAJORITY = "no_majority"
+
+
+class SynchronizedData(BaseSynchronizedData):
+    """
+    Class to represent the synchronized data.
+
+    This data is replicated by the tendermint application.
+    """
+
+    @property
+    def mech_requests(self) -> list:
+        """Get the mech requests."""
+        return cast(list, self.db.get("mech_requests", []))
+
+
+    @property
+    def final_tx_hash(self) -> str:
+        """Get the verified tx hash."""
+        return cast(str, self.db.get_strict("final_tx_hash"))
+
+
+class MechRequestRound(OnlyKeeperSendsRound):
+    """MechRequestRound"""
+
+    payload_class = MechRequestPayload
+    synchronized_data_class = SynchronizedData
+
+    def end_block(self) -> Optional[Tuple[BaseSynchronizedData, Event]]:
+        """Process the end of the block."""
+        if self.keeper_payload is None:
+            return None
+
+        mech_requests = cast(
+            SynchronizedData, self.synchronized_data
+        ).mech_requests
+
+        # Assing the request hash to its corresponding requests
+        for r in mech_requests:
+            if "tx_hash" not in r:
+                r["tx_hash"] = cast(MechRequestPayload, self.keeper_payload).tx_hash  # FIXME: do multi-requests have 1 hash only?
```

Comment:
Yes they have one hash, the corresponding request ids will be more than 1 though. Check the comments above re the required contract change.